rllib/agents/trainer.py

```diff
@@ -980,20 +1126,28 @@ def auto_duration_fn(unit, num_eval_workers, eval_cfg, num_units_done):
 
         # No evaluation necessary, just run the next training iteration.
         if not evaluate_this_iter:
-            step_results = next(self.train_exec_impl)
+            if use_exec_api:
+                step_results = next(self.train_exec_impl)
+            else:
+                step_results = self.training_iteration()
         # We have to evaluate in this training iteration.
         else:
             # No parallelism.
             if not self.config["evaluation_parallel_to_training"]:
-                step_results = next(self.train_exec_impl)
+                if use_exec_api:
+                    step_results = next(self.train_exec_impl)
+                else:
+                    step_results = self.training_iteration()
 
             # Kick off evaluation-loop (and parallel train() call,
             # if requested).
             # Parallel eval + training.
             if self.config["evaluation_parallel_to_training"]:
                 with concurrent.futures.ThreadPoolExecutor() as executor:
                     train_future = executor.submit(
-                        lambda: next(self.train_exec_impl))
+                        lambda: (next(self.train_exec_impl) if
+                                 use_exec_api else
+                                 self.training_iteration()))
```

Comment:
the same if-else is repeated 3 time here. make a local helper func? and you can simply submit it here.

Reply:
Great catch! Done.